llvm/lib/Transforms/IPO/ExpandVariadics.cpp

```diff
@@ -0,0 +1,1056 @@
+  Builder.SetInsertPointPastAllocas(CBF);
+
+  // The struct instance needs to be at least MaxFieldAlign for the alignment of
+  // the fields to be correct at runtime. Use the native stack alignment instead
+  // if that's greater as that tends to give better codegen.
+  Align AllocaAlign = MaxFieldAlign;
+  if (DL.exceedsNaturalStackAlignment(Align(1024))) {
+    // TODO: DL.getStackAlignment could return a MaybeAlign instead of assert
+    AllocaAlign = std::max(AllocaAlign, DL.getStackAlignment());
+  }
+
+  AllocaInst *Alloced = Builder.Insert(
+      new AllocaInst(VarargsTy, DL.getAllocaAddrSpace(), nullptr, AllocaAlign),
+      "vararg_buffer");
+  Changed = true;
+  assert(Alloced->getAllocatedType() == VarargsTy);
+
+  // Initialise the fields in the struct
+  Builder.SetInsertPoint(CB);
+
+  Builder.CreateLifetimeStart(Alloced, sizeOfAlloca(Ctx, DL, Alloced));
+
+  Frame.initialiseStructAlloca(DL, Builder, Alloced);
+
+  unsigned NumArgs = FuncType->getNumParams();
+
+  SmallVector<Value *> Args;
+  Args.assign(CB->arg_begin(), CB->arg_begin() + NumArgs);
+
+  // Initialise a va_list pointing to that struct and pass it as the last
+  // argument
+  AllocaInst *VaList = nullptr;
+  {
+    if (!ABI.VAList->passedInSSARegister()) {
+      Type *VaListTy = ABI.VAList->vaListType(Ctx);
+      Builder.SetInsertPointPastAllocas(CBF);
+      VaList = Builder.CreateAlloca(VaListTy, nullptr, "va_list");
+      Builder.SetInsertPoint(CB);
+      Builder.CreateLifetimeStart(VaList, sizeOfAlloca(Ctx, DL, VaList));
+    }
+    Args.push_back(ABI.VAList->initializeVAList(Ctx, Builder, VaList, Alloced));
+  }
+
+  // Attributes excluding any on the vararg arguments
+  AttributeList PAL = CB->getAttributes();
+  if (!PAL.isEmpty()) {
+    SmallVector<AttributeSet, 8> ArgAttrs;
+    for (unsigned ArgNo = 0; ArgNo < NumArgs; ArgNo++)
+      ArgAttrs.push_back(PAL.getParamAttrs(ArgNo));
+    PAL =
+        AttributeList::get(Ctx, PAL.getFnAttrs(), PAL.getRetAttrs(), ArgAttrs);
+  }
+
+  SmallVector<OperandBundleDef, 1> OpBundles;
+  CB->getOperandBundlesAsDefs(OpBundles);
+
+  CallBase *NewCB = nullptr;
+  // TODO, other instructions? Haven't managed to write variadic inline asm yet
+  if (CallInst *CI = dyn_cast<CallInst>(CB)) {
+
+    Value *Dst = NF ? NF : CI->getCalledOperand();
+    FunctionType *NFTy = inlinableVariadicFunctionType(M, VarargFunctionType);
+
+    NewCB = CallInst::Create(NFTy, Dst, Args, OpBundles, "", CI);
+
+    CallInst::TailCallKind TCK = CI->getTailCallKind();
+    assert(TCK != CallInst::TCK_MustTail); // guarded at prologue
+
+    // It doesn't get to be a tail call any more
+    // might want to guard this with arch, x64 and aarch64 document that
+    // varargs can't be tail called anyway
+    // Not totally convinced this is necessary but dead store elimination
+    // will discard the stores to the Alloca and pass uninitialised memory along
+    // instead when the function is marked tailcall
```

Comment:
I'll take another shot at explaining this...

From LangRef: "[tail and musttail] imply that the callee does not access allocas from the caller."  This applies whether or not it's actually possible to do sibling-call/tail-call optimization on the call.  And in fact, "tail" has basically no other semantic implications.  So LLVM IR optimizations infer "tail" pretty aggressively.

Because of this definition, alias analysis assumes calls marked "tail" don't read/write allocas. So the DSE behavior is expected.

Reply:
Tail on varargs functions is sort of valid in that the callee doesn't access an alloca. The callee does tend to dig around in the stack of the caller though, so tail calling varargs functions only works out nicely if the callee rearranges the stack frame on behalf of the caller. Which it can do, but usually doesn't.

I think your reading is right that DSE making this inference is reasonable. I think it's also right that a varargs function could be tail called on some implementations, though isn't on the ones considered so far in this patch.